### PR TITLE
HOCS-2986: remove hardcoded emails

### PR DESCRIPTION
**Pull request by LiamMacP**
Currently the whitelisted domains that can be used for user 
additions are hardcoded within `application.properties`. This 
change removes this and allows it to be pulled in through a new 
config map.

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidator.java [hunk 1]
```diff
@@ -12,7 +13,7 @@
 public class EmailDomainValidator implements ConstraintValidator<EmailDomainWhitelisted, String> {
 
     @Value("${user.email.whitelist}")
-    private List<String> whitelistedDomains = asList("homeoffice.gov.uk");
+    private List<String> whitelistedDomains = List.of("homeoffice.gov.uk");
```
Comment (zuzak) [inline src/main/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidator.java:16]: Will this default to homeoffice.gov.uk if the config var is not set? I wonder if either "disabling user creation if domains not given" or "allowing any email" would be more sensible defaults?

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -57,7 +57,7 @@ auditing.deployment.name=hocs-info-service
 aws.sqs.access.key=12345
 audit.aws.sqs.access.key=12345
 
-user.email.whitelist=homeoffice.gov.uk,hmpo.gov.uk,cluster2security.gov.uk
+user.email.whitelist=${user.addition.allowed.domains:}
```
Comment (zuzak) [inline src/main/resources/application.properties:60]: This naming makes it seem like it's adding to the existing list of allowed domains, but it's actually overwriting it. Something like `$ALLOWED_EMAIL_DOMAINS_FOR_USER_CREATION` or something idk?

File: src/test/java/uk/gov/digital/ho/hocs/info/api/UserResourceRestTest.java [hunk 3]
```diff
@@ -32,7 +32,7 @@
 
   @Test
   public void testCreateUserWithWhitelistedDomains() throws Exception {
-    List<String> whitelistedDomains = List.of("mick@homeoffice.gov.uk", "mick@hmpo.gov.uk");
+    List<String> whitelistedDomains = List.of("mick@example.co.uk", "mick@example.org");
```
Comment (zuzak) [inline src/test/java/uk/gov/digital/ho/hocs/info/api/UserResourceRestTest.java:35]: Incidentally might be nice to have some tests that check if subdomains are let through

Comment (LiamMacP) [reply src/test/java/uk/gov/digital/ho/hocs/info/api/UserResourceRestTest.java:35]: There is a new test file added that tests the validator more explicitely. This contains a test for the subdomains to ensure they are not part of this.

File: src/main/resources/application.properties (hunk 2 above)
Comment (LiamMacP) [reply src/main/resources/application.properties]: Change to that - much better - will do the corresponding change to the kube repo.

File: src/main/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidator.java (hunk 1 above)
Comment (LiamMacP) [reply src/main/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidator.java]: Changed to default to an empty list. This disables when domains are not provided.

File: src/test/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        assertThat(emailDomainValidator.isValid("test@test.com", constraintValidatorContext))
+                .isFalse();
+    }
+
+    @Test
+    public void whenSubdomainEmailAddressIsValidButNotInList_ShouldReturnFalse(){
+        assertThat(emailDomainValidator.isValid("test@sub.example.com", constraintValidatorContext))
+                .isFalse();
+    }
+
```
Comment (zuzak) [inline src/test/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidatorTest.java:57]:
```suggestion

    @Test
    public void whenPrefixedDomainEmailAddressIsValidButNotInList_ShouldReturnFalse(){
        assertThat(emailDomainValidator.isValid("test@notexample.com", constraintValidatorContext))
                .isFalse();
    }
```

**Review by zuzak — COMMENTED** [review]
Can we get a test for behaviour when `user.email.whitelist` is not defined?

File: src/main/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidator.java [hunk 5]
```diff
@@ -12,13 +13,17 @@
...
     public void initialize(EmailDomainWhitelisted emailDomainWhitelisted) { }
 
     @Override
     public boolean isValid(String emailAddress, ConstraintValidatorContext cxt) {
-      return emailAddress == null || emailAddress.isBlank() || whitelistedDomains.stream().anyMatch(emailAddress::endsWith);
+        return emailAddress != null
+                && !emailAddress.isBlank()
+                && whitelistedDomains
+                    .stream()
+                    .anyMatch(domain -> emailAddress.toLowerCase().endsWith(String.format("@%s", domain)));
```
Comment (zuzak) [inline src/main/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidator.java:27]: While `foo@example.com` and `foo@EXAMPLE.COM` are the same email address, according to RFC 5322  `foo@example.com` and `FOO@example.com` aren't -- although almost all mailservers these days will treat them as such.

If there's a handy `normalizeEmail` method handy that could be better than a simple`.toLowerCase()`.

**Review by Paulcr-ho — APPROVED** [review]

File: src/main/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidator.java (hunk 5 above)
Comment (Paulcr-ho) [reply src/main/java/uk/gov/digital/ho/hocs/info/utils/EmailDomainValidator.java:27]: toLower() is sufficient in this situation.

**Comment by LiamMacP** [comment]
> Can we get a test for behaviour when `user.email.whitelist` is not defined?

`user.email.whitelist` is always defined in some form - whether it be a list from the configmap or empty due to the substitution.

**Pull request merged**